c_src/adbc_nif.cpp

```diff
@@ -277,6 +277,95 @@ static ERL_NIF_TERM get_arrow_array_list_children(ErlNifEnv *env, struct ArrowSc
     return enif_make_list_from_array(env, children.data(), (unsigned)items_values->n_children);
 }
 
+static ERL_NIF_TERM unix_second_offset_to_date(ErlNifEnv *env, uint64_t seconds) {
+    time_t t = (time_t)seconds;
+    tm* time = gmtime(&t);
+
+    ERL_NIF_TERM date;
+    ERL_NIF_TERM keys[] = {
+        erlang::nif::atom(env, "__struct__"),
+        erlang::nif::atom(env, "calendar"),
+        erlang::nif::atom(env, "year"),
+        erlang::nif::atom(env, "month"),
+        erlang::nif::atom(env, "day")
+    };
+    ERL_NIF_TERM values[] = {
+        erlang::nif::atom(env, "Elixir.Date"),
+        erlang::nif::atom(env, "Elixir.Calendar.ISO"),
+        enif_make_int(env, time->tm_year + 1900),
+        enif_make_int(env, time->tm_mon + 1),
+        enif_make_int(env, time->tm_mday)
+    };
+    enif_make_map_from_arrays(env, keys, values, sizeof(keys)/sizeof(keys[0]), &date);
+    return date;
+}
+
+static ERL_NIF_TERM nanoseconds_to_time(ErlNifEnv *env, uint64_t ns) {
+    // Elixir only supports microsecond precision
+    uint64_t us = ns / 1000;
+    time_t s = (time_t)(us / 1000000);
+    tm* time = gmtime(&s);
+
+    us = us % 1000000;
+    uint64_t us_precision = 6;
+
+    ERL_NIF_TERM t;
+    ERL_NIF_TERM keys[] = {
+        erlang::nif::atom(env, "__struct__"),
+        erlang::nif::atom(env, "calendar"),
+        erlang::nif::atom(env, "hour"),
+        erlang::nif::atom(env, "minute"),
+        erlang::nif::atom(env, "second"),
+        erlang::nif::atom(env, "microsecond")
+    };
+    ERL_NIF_TERM values[] = {
+        erlang::nif::atom(env, "Elixir.Time"),
+        erlang::nif::atom(env, "Elixir.Calendar.ISO"),
+        enif_make_int(env, time->tm_hour),
+        enif_make_int(env, time->tm_min),
+        enif_make_int(env, time->tm_sec),
+        enif_make_tuple2(env, enif_make_int(env, us), enif_make_int(env, us_precision))
+    };
+    enif_make_map_from_arrays(env, keys, values, sizeof(keys)/sizeof(keys[0]), &t);
+    return t;
+}
+
+static ERL_NIF_TERM unix_timestamp_to_naive_datetime(ErlNifEnv *env, uint64_t timestamp) {
+    // Elixir only supports microsecond precision
+    uint64_t us = timestamp / 1000;
+    time_t t = (time_t)(us / 1000000);
+    tm* time = gmtime(&t);
+
+    us = us % 1000000;
+    uint64_t us_precision = 6;
```

Comment:
Is there a better way to handle microsecond precision than just always assuming 6? Snowflake always uses the `ttn` (time64 nanoseconds) format, even if the source data doesn't have nanosecond precision, e.g. these two return the same value:

```sql
select '00:00:00.000000';
select '00:00:00';
```

Reply:
When you compute the unit, you can compute the precision too! s=0, ms=3, us=6. :) But the first tuple element is always in microseconds.